modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,764 @@
+//! if required.
+
+use crate::BridgeStorage;
+use bp_substrate::{prove_finality, AuthoritySet, ImportedHeader, ScheduledChange};
+use sp_finality_grandpa::{ConsensusLog, GRANDPA_ENGINE_ID};
+use sp_runtime::generic::OpaqueDigestItemId;
+use sp_runtime::traits::{CheckedAdd, Header as HeaderT, One};
+use sp_std::{prelude::Vec, vec};
+
+/// The finality proof used by the pallet.
+///
+/// For a Substrate based chain using Grandpa this will
+/// be an encoded Grandpa Justification.
+pub type FinalityProof<'a> = &'a [u8];
+
+/// Errors which can happen while importing a header.
+#[derive(Debug, PartialEq)]
+pub enum ImportError {
+	/// This header is older than our latest finalized block, thus not useful.
+	OldHeader,
+	/// This header has already been imported by the pallet.
+	HeaderAlreadyExists,
+	/// We're missing a parent for this header.
+	MissingParent,
+	/// The number of the header does not follow its parent's number.
+	InvalidChildNumber,
+	/// The height of the next authority set change overflowed.
+	ScheduledHeightOverflow,
+}
+
+/// Errors which can happen while verifying a headers finality.
+#[derive(Debug, PartialEq)]
+pub enum FinalizationError {
+	/// This header has never been imported by the pallet.
+	UnknownHeader,
+	/// We were unable to prove finality for this header.
+	UnfinalizedHeader,
+	/// Trying to prematurely import a justification
+	PrematureJustification,
+	/// We failed to verify this header's ancestry.
+	AncestryCheckFailed,
+	/// This header is older than our latest finalized block, thus not useful.
+	OldHeader,
+}
+
+/// Used to verify imported headers and their finality status.
+#[derive(Debug)]
+pub struct Verifier<S> {
+	pub storage: S,
+}
+
+impl<S, H> Verifier<S>
+where
+	S: BridgeStorage<Header = H>,
+	H: HeaderT,
+{
+	/// Import a header to the pallet.
+	///
+	/// Will perform some basic checks to make sure that this header doesn't break any assumptions
+	/// such as being on a different finalized fork.
+	pub fn import_header(&mut self, header: H) -> Result<(), ImportError> {
+		let best_finalized = self.storage.best_finalized_header();
+
+		if header.number() <= best_finalized.number() {
+			return Err(ImportError::OldHeader);
+		}
+
+		if self.storage.header_exists(header.hash()) {
+			return Err(ImportError::HeaderAlreadyExists);
+		}
+
+		let parent_header = self
+			.storage
+			.header_by_hash(*header.parent_hash())
+			.ok_or(ImportError::MissingParent)?;
+
+		let parent_number = *parent_header.number();
+		if parent_number + One::one() != *header.number() {
+			return Err(ImportError::InvalidChildNumber);
+		}
+
+		// This header requires a justification since it enacts an authority set change. We don't
+		// need to act on it right away (we'll update the set once this header gets finalized), but
+		// we need to make a note of it.
+		//
+		// Note: This assumes that we can only have one authority set change pending at a time.
```

Comment:
Just asking - we're going to fix it in next PRs, right? I'd replace 'Note' with 'TODO' here, even without submitting an issue (because pallet is in-progress anyway).